### DISCUSSION TITLE
buildTarget/restore

### DISCUSSION THREAD

**Discussion by schinnerld** [build-server-protocol/build-server-protocol#675] [Q&A]
I was wondering if there should be an explicit buildTarget/restore request. Currently BSP has no direct way to handle dependency restore (like dotnet restore). 

In your opinion, would this be useful? If not, is there a recommended way to call dependency restore (e.g. with buildTarget/compile arguments)? Or do you think it should just be an implicit step (task) of the compile request?

**Comment by jastice** [comment]
Hi! Can you describe what the semantics of restore should be and why it is needed? I'm not familiar with dotnet restore.

**Comment by agluszak** [comment]
I also don't know what you mean by "dependency restore"

**Comment by schinnerld** [comment]
Sorry for the confusion. I'm talking about package managers to download and update necessary libraries. In the .NET ecosystem for example that would be NuGet.

The dotnet commands can be wonderfully matched to the BSP requests. 
[dotnet build](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-build) = buildTarget/compile
[dotnet run](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-run) = buildTarget/run
[dotnet test ](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-test)= builtTarget/test
[dotnet clean](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-clean) = buildTarget/cleanCache

But I miss a [dotnet restore](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-restore) counterpart, hence my question.

**Reply by tgodzik** [reply]
Servers like the Bazel one would get the dependencies at the start since that is anyway needed to compile anything. We also have a reload request that could work here for updating deps etc.

**Reply by adpi2** [reply]
It seems the .NET documentation of [restore](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-restore#description) and @tgodzik say about the same thing.

> In most cases, you don't need to explicitly use the dotnet restore command, since if a NuGet restore is necessary, the following commands run it implicitly:
dotnet build
dotnet run
dotnet test
...

So maybe we don't need such a method yet. Or maybe you can use the [buildTarget/dependencyModules](https://build-server-protocol.github.io/docs/specification#buildtargetdependencymodules-request) request instead, which needs to resolve and download the dependencies to return them to the client.
